### ISSUE TITLE
Revamp project handling UI

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#827]
The current project handling UI grew organizally from whatever happened to already be there, and people find it confusing.  It should be replaced by something that acts more intuitively.  Some improvements would be:

1. A tree view whose parts can be expanded and collapsed at will, instead of having only one currently expanded path.
2. Reordering and moving projects by drag and drop (with perhaps a sort button that puts the projects back in order, by time or alphabetically?).
3. A search or filter bar to look for projects based on substrings of the name.

All of this fits into the current model (with some added metadata for ordering).  It's just a client-side revamp.

**Comment by nixorn** [comment]
I going to try. Looks, need to configure local auth for this.

**Comment by nixorn** [comment]
What I see is we have several different variables which represents current folder tree. It contains `window.allProjectNames`, `window.allFolderNames` and `window.nestedDirs`. It looks over-complicated.

Together these variables represent what user currently see, but hidden parts of tree have no representation. There is no single structure which reflects all user directory tree with all projects.

I guess this structure could be something like

```javascript
[
  { type: "directory",
    name: "folder1",
    entries: [
      { type: "directory",
        name: "folder2",
        entries: [{type: project, name: "project1"}, {type:"project", name: {project2"}]
      }]
  }
]
```

**Comment by cdsmith** [comment]
That sounds good to me, sure.  So step 1 here could be to switch the representations, but leave the UI the same.  This needs to be done very carefully, because there's a lot of code that depends on the existing representation, and it would be easy to break a bunch of stuff.

**Comment by nixorn** [comment]
Ok, I'll go this way.

**Comment by nixorn** [comment]
Project paths have next structure `DWh/DWhBei51A4TKXgNYuoiZdig/DrQ/DrQI0gpIFuQMxlrqBj3qHKw/SMa/SMaLGg9zLg--LjR7kMpDfYg.cw`. For what purpose is these 3-symbols prefix dirs `DWh`, `SMa`?

**Comment by nixorn** [comment]
What is the history of this decision?

**Comment by cdsmith** [comment]
These three-letter prefixes originally came about because directories were getting too large.  They are certainly overkill for a single user's project directory, since you're going to load them all anyway.  I didn't notice when the original folder-structure pull request was sent.

I'd be okay with doing a migration if you want to fix this.  The best way to do this is probably to add the migration code to the beginning of codeworld-server.  You'd need to fix all the server code to work with the new path structure at the same time.

**Comment by nixorn** [comment]
Yes, I want fix this. I'm not sure what means "to the beginning of codeworld-server".
I think migration should be run once with logic update, between old server down and new server up. Maybe as different binary file or even bash script.

**Comment by cdsmith** [comment]
If I were running the only server, then a one-time migration script would be fine.  But that's not the case.  I wouldn't want us to end up in a situation where someone else runs `git pull` and rebuilds, and suddenly their server is broken.

If you'd like to write a separate script for the migration, that's okay; but it needs to be added to `run.sh` so we don't have to ask everyone hosting a server to take an extra step, and therefore it should be okay to run every time the server is started.  (For example, it can check if the migration has already happened, and exit right away if so.)

**Comment by nixorn** [comment]
Yes, got it.

**Comment by cdsmith** [comment]
I'm reopening this, because even though the implementation now uses a tree view, the UI still looks like a stack of buttons just as it always did.  So this didn't really fix the root issue.
